src/cysignals/implementation.c

```diff
@@ -107,6 +108,60 @@ static inline void reset_CPU(void)
 }
 
 
+/* Reset all signal handlers and the signal mask to their defaults. */
+static inline void sig_reset_defaults(void) {
+    signal(SIGHUP, SIG_DFL);
+    signal(SIGINT, SIG_DFL);
+    signal(SIGQUIT, SIG_DFL);
+    signal(SIGILL, SIG_DFL);
+    signal(SIGABRT, SIG_DFL);
+    signal(SIGFPE, SIG_DFL);
+    signal(SIGBUS, SIG_DFL);
+    signal(SIGSEGV, SIG_DFL);
+    signal(SIGALRM, SIG_DFL);
+    signal(SIGTERM, SIG_DFL);
+    sigprocmask(SIG_SETMASK, &default_sigmask, NULL);
+}
+
+
+/* Call sigdie() with the appropriate error message for the signal, or
+ * whether the exception occurred inside our signal handler */
+static inline void sigdie_for_sig(int sig, int inside)
+{
+    if (inside) sigdie(sig, "An error occurred during signal handling.");
+
+    /* Quit Python with an appropriate message. */
+    switch(sig)
+    {
+        case SIGQUIT:
+            sigdie(sig, NULL);
+            break;  /* This will not be reached */
+        case SIGILL:
+            sigdie(sig, "Unhandled SIGILL: An illegal instruction occurred.");
+            break;  /* This will not be reached */
+        case SIGABRT:
+            sigdie(sig, "Unhandled SIGABRT: An abort() occurred.");
+            break;  /* This will not be reached */
+        case SIGFPE:
+            sigdie(sig, "Unhandled SIGFPE: An unhandled floating point exception occurred.");
+            break;  /* This will not be reached */
+        case SIGBUS:
+            sigdie(sig, "Unhandled SIGBUS: A bus error occurred.");
+            break;  /* This will not be reached */
+        case SIGSEGV:
+            sigdie(sig, "Unhandled SIGSEGV: A segmentation fault occurred.");
+            break;  /* This will not be reached */
+    };
+    sigdie(sig, "Unknown signal received.\n");
+}
+
+
+/* Additional platform-specific implementation code */
+#if defined(__CYGWIN__) && defined(__x86_64__)
+#include "implementation_cygwin.h"
```

Comment:
I find an `#include` in the middle of the code bad style. Can you put it near the top of the file?

Reply:
I agree, though though the file's contents depends on some inline functions I added above.  Without it I get complaints about missing declarations--in that case it might almost be better to add an `implementation.h` and move the declarations there.

Reply:
> Without it I get complaints about missing declarations

Just add declarations in the beginning of `implementation.c`.

> in that case it might almost be better to add an implementation.h

My feeling is that header files are mostly used to define a public API, which is not the case here.

Reply:
That's not necessarily true at all--headers are just so that declarations an be shared between compilation units, and doesn't necessarily have anything to do with public APIs.  It's not unusual to have headers that are used internally for build purposes, but are never "installed".